crates/biome_analyze/src/lib.rs

```diff
@@ -581,136 +528,26 @@ where
-                        suppressed_instances.insert(value.to_owned(), key);
-                        if suppression.is_top_level() {
-                            self.top_level_suppressions.insert(key);
-                        }
-                        if self.top_level_suppressions.has_filter(&key) {
-                            already_suppressed = true
-                        }
-                    }
-                    (Some(key), None) => {
-                        if suppression.is_top_level() {
-                            self.top_level_suppressions.insert(key);
-                        }
-                        if self.top_level_suppressions.has_filter(&key) {
-                            already_suppressed = true
-                        }
-                        suppressed_rules.insert(key);
-                    }
-                    _ if range_match(self.range, range) => {
-                        // Emit a warning for the unknown rule
-                        let signal = DiagnosticSignal::new(move || match group_rule {
-                            Some((group, rule)) => AnalyzerSuppressionDiagnostic::new(
-                                category!("suppressions/unknownRule"),
-                                range,
-                                format_args!(
-                                    "Unknown lint rule {group}/{rule} in suppression comment"
-                                ),
-                            ),
-
-                            None => AnalyzerSuppressionDiagnostic::new(
-                                category!("suppressions/unknownGroup"),
-                                range,
-                                format_args!(
-                                    "Unknown lint rule group {rule} in suppression comment"
-                                ),
-                            ),
-                        });
-
-                        (self.emit_signal)(&signal)?;
-                    }
-                    _ => {}
-                }
-            } else {
-                suppressed_rules.clear();
-                suppress_all = true;
-                // If this if a "suppress all lints" comment, no need to
-                // parse anything else
-                break;
+            if let Err(diagnostic) =
+                self.suppressions
+                    .push_suppression(&suppression, range, token.text_range())
+            {
+                let signal = DiagnosticSignal::new(|| diagnostic.clone());
+                (self.emit_signal)(&signal)?;
+                continue;
             }
-        }
-
-        if is_top_level_suppression && !misplaced {
-            self.top_level_suppressions.expand_range(range);
-            return ControlFlow::Continue(());
-        }
 
-        if !suppress_all && suppressed_rules.is_empty() && suppressed_instances.is_empty() {
-            return ControlFlow::Continue(());
+            has_suppressions |= true;
```

Comment:
```suggestion
            has_suppressions = true;
```

(The effect’s the same, but the intention is clearer.)